tools/clang/unittests/HLSLExec/ExecutionTest_SER.h

```diff
@@ -0,0 +1,2554 @@
+    float4 U;
+    float4 V;
+    float4 W;
+    float sceneScale;
+    uint2 WindowSize;
+    int rayFlags;    
+};
+
+struct CustomAttrs
+{
+    float dist;
+};
+
+struct[raypayload] PerRayData
+{
+    uint visited : read(anyhit, closesthit, miss, caller) : write(anyhit, miss, closesthit, caller);
+};
+
+// reordercoherent // Requires #7250
+RWStructuredBuffer<int> testBuffer : register(u0);
+RaytracingAccelerationStructure topObject : register(t0);
+ConstantBuffer<SceneConstants> sceneConstants : register(b0);
+
+RayDesc ComputeRay()
+{
+    uint2 launchIndex = DispatchRaysIndex().xy;
+    uint2 launchDim = DispatchRaysDimensions().xy;
+
+    float2 d = float2(DispatchRaysIndex().xy) / float2(DispatchRaysDimensions().xy) * 2.0f - 1.0f;
+    RayDesc ray;
+    ray.Origin = sceneConstants.eye.xyz;
+    ray.Direction = normalize(d.x * sceneConstants.U.xyz + d.y * sceneConstants.V.xyz + sceneConstants.W.xyz);
+    ray.TMin = 0;
+    ray.TMax = 1e18;
+
+    return ray;
+}
+
+[shader("raygeneration")]
+void raygen()
+{
+    uint2 launchIndex = DispatchRaysIndex().xy;
+    uint2 launchDim = DispatchRaysDimensions().xy;
+
+    RayDesc ray = ComputeRay();
+
+    PerRayData payload;
+    payload.visited = 0;
+
+    dx::HitObject hitObject = dx::HitObject::TraceRay(topObject, RAY_FLAG_NONE, 0xFF, 0, 1, 0, ray, payload);
+    dx::MaybeReorderThread(hitObject);
+    
+    // Check Attributes for hit detection.
+    CustomAttrs customAttrs = hitObject.GetAttributes<CustomAttrs>();
+    bool isHit = hitObject.IsHit();
+
+    int testVal = 0;
+    if (isHit) {
+        if (int(floor(customAttrs.dist)) % 2 == 0)
+          testVal = hitObject.GetHitKind();
+    }
+    else
+    {
+        // Use 255 to keep outside the HitKind range [0, 127] we passthru for hits.
+        testVal = 255;
+    }
+    int id = launchIndex.x + launchIndex.y * launchDim.x;
+    testBuffer[id] = testVal;
+}
+
+[shader("miss")]
+void miss(inout PerRayData payload)
+{
+    // UNUSED
+}
+
+[shader("anyhit")]
+void anyhit(inout PerRayData payload, in CustomAttrs attrs)
+{
+    AcceptHitAndEndSearch();
+}
+
+[shader("closesthit")]
+void closesthit(inout PerRayData payload, in CustomAttrs attrs)
+{
+    // UNUSED
+}
+
+[shader("intersection")]
+void intersection()
+{
+    // hitPos is intersection point with plane (base, n)
+    float3 base = {0.0f,0.0f,0.5f};
+    float3 n = normalize(float3(0.0f,0.5f,0.5f));
+    float t = dot(n, base - ObjectRayOrigin()) / dot(n, ObjectRayDirection());
+    if (t > RayTCurrent() || t < RayTMin()) {
+        return;
+    }
+    float3 hitPos = ObjectRayOrigin() + t * ObjectRayDirection();
+    float3 relHitPos = hitPos - base;
+    // Encode some hit information in hitKind
+    int hitKind = 0;
+    if (relHitPos.y >= 0.0f)
+        hitKind = 1;
+    hitKind *= 2;
+    if (relHitPos.x >= 0.0f)
+        hitKind += 1;
+    hitKind *= 2;
+    if (relHitPos.z >= 0.0f)
+        hitKind += 1;
+
+    CustomAttrs attrs;
+    attrs.dist = length(relHitPos);
```

Comment:
It seems something in here is too dependent on floating point precision. If I allow FMA fusing in WARP (which literally just was hooked up a few weeks ago) then the test passes, but if I disable it then I get a bucketing which is off-by-one from the test expectations.

Reply:
Made the hitkind more robust by replacing the fp comparison and with launchIndex-based comparisons to generate a meaningful hitKind. Please test on your end, there may still be fp precision issues with the aabb intersection logic itself.